WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java

```diff
@@ -0,0 +1,116 @@
+package org.wordpress.android.util;
+
+import android.util.Base64;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.libsodium.jni.NaCl;
+
+public class EncryptionUtils {
+    static final int BOX_SEALBYTES = NaCl.sodium().crypto_box_sealbytes();
+    static final int XCHACHA20POLY1305_KEYBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_keybytes();
+    static final int XCHACHA20POLY1305_STATEBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_statebytes();
+    static final int XCHACHA20POLY1305_HEADERBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_headerbytes();
+    static final int XCHACHA20POLY1305_ABYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_abytes();
+    static final int BASE64_ENCODE_FLAGS = Base64.DEFAULT;
+
+    static final short XCHACHA20POLY1305_TAG_FINAL = 
+            (short) NaCl.sodium().crypto_secretstream_xchacha20poly1305_tag_final();
+    static final short XCHACHA20POLY1305_TAG_0 = (short) 0;
+
+    static final String KEYED_WITH = "v1";
+
+    /*
+        Returns a JSON String containing following data:
+        {
+            "keyedWith": "v1",
+            "encryptedKey": "$key_as_base_64",  // The encrypted AES key
+            "header": "base_64_encoded_header", // The xchacha20poly1305 stream header
+            "messages": []                      // the stream elements, base-64 encoded
+        }
+    */
+    public static String encryptStringData(final String publicKeyBase64,
+                                           final String stringData,
+                                           final String delimiter) throws JSONException {
+        JSONObject encryptionDataJson = new JSONObject();
+        encryptionDataJson.put("keyedWith", KEYED_WITH);
+
+        // Create data-specific key
+        byte[] key = new byte[XCHACHA20POLY1305_KEYBYTES];
+        NaCl.sodium().crypto_secretstream_xchacha20poly1305_keygen(key);
+
+        final String encryptedKeyBase64 = getBoxSealEncryptedBase64String(
+                publicKeyBase64,
+                key,
+                XCHACHA20POLY1305_KEYBYTES);
+        encryptionDataJson.put("encryptedKey", encryptedKeyBase64);
+
+        byte[] state = new byte[XCHACHA20POLY1305_STATEBYTES];
+
+        final String headerBase64 = initSecretStreamXchacha20poly1305(state, key);
+        encryptionDataJson.put("header", headerBase64);
+
+        String[] splitStringData = stringData.split(delimiter);
+        JSONArray encryptedElementsJson = new JSONArray();
+  
+
+        for (String element : splitStringData) {
+            element = element + delimiter; // Add delimiter back to the end of the line
```

Comment:
This changes the content of the message for the last line, which we probably shouldn't do :)

Reply:
Oh boy.  Thank you for catching that. 😅  Yeesh.  😕

Reply:
fixed the bug here: 09c15a87bc1be23431b4f78d532b5b72837d780c

Reply:
oops. through rewriting my test case I figured out that this actually wasn't a bug. 🙂 unfixed it here: aa875f161ce9cb2c2745b65a0181db1eb516005f